app/models/voxelytics/LokiClient.scala

```diff
@@ -0,0 +1,173 @@
+          case e =>
+            logger.error(s"Unexpected error $e")
+            Fox.failure("Unexpected error while trying to connect to Loki.", Full(e))
+        })
+      isServerAvailable <- isServerAvailableBox.toFox
+      _ <- if (!isServerAvailable) {
+        waitAndRecurse(until)
+      } else {
+        logger.info("Loki is available.")
+        Fox.successful(())
+      }
+    } yield ()
+  }
+
+  def queryLogs(runName: String,
+                organizationId: ObjectId,
+                taskName: Option[String],
+                minLevel: VoxelyticsLogLevel = VoxelyticsLogLevel.INFO,
+                startTime: Instant,
+                endTime: Instant,
+                limit: Long,
+                backward: Boolean = true): Fox[JsValue] = {
+    val levels = VoxelyticsLogLevel.sortedValues.drop(VoxelyticsLogLevel.sortedValues.indexOf(minLevel))
+
+    val logQLFilter = List(
+      taskName.map(t => s"""vx_task_name="$t""""),
+      Some(s"""level=~"(${levels.mkString("|")})"""")
+    ).flatten.mkString(" | ")
+    val logQL =
+      s"""{vx_run_name="$runName",wk_org="${organizationId.id}",wk_url="${wkConf.Http.uri}"} | json vx_task_name,level | $logQLFilter"""
+
+    val queryString =
+      List("query" -> logQL,
+           "start" -> startTime.toString,
+           "end" -> endTime.toString,
+           "limit" -> limit.toString,
+           "direction" -> (if (backward) {
+                             "backward"
+                           } else {
+                             "forward"
+                           })).map(kv => s"${kv._1}=${java.net.URLEncoder.encode(kv._2, "UTF-8")}").mkString("&")
+    for {
+      _ <- serverStartupFuture
+      res <- rpc(s"${conf.uri}/loki/api/v1/query_range?$queryString").silent.getWithJsonResponse[JsValue]
+    } yield
+      JsArray(
+        (res \ "data" \ "result")
+          .as[List[JsValue]]
+          .flatMap(
+            stream =>
+              (stream \ "values")
+                .as[List[(String, String)]]
+                .map(value =>
+                  Json.parse(value._2).as[JsObject] ++ (stream \ "stream").as[JsObject] ++ Json.obj(
+                    "timestamp" -> Instant(value._1.substring(0, value._1.length - 6).toLong)))))
+  }
+
+  def bulkInsert(logEntries: List[JsValue], organizationId: ObjectId): Fox[Unit] =
+    if (logEntries.nonEmpty) {
+      val streams = logEntries
+        .groupBy(
+          entry => ((entry \ "vx" \ "workflow_hash").as[String], (entry \ "vx" \ "run_name").as[String])
+        )
+        .map(kv =>
+          Json.obj(
+            "stream" -> Json.obj("vx_workflow_hash" -> kv._1._1,
+                                 "vx_run_name" -> kv._1._2,
+                                 "wk_url" -> wkConf.Http.uri,
+                                 "wk_org" -> organizationId.id),
+            "values" -> JsArray(kv._2.map(entry => {
+              val timestamp = java.time.Instant.parse((entry \ "@timestamp").as[String]).toEpochMilli
```

Comment:
Could this use our Instant class as well? please add the missing functions like toNanoseconds